grpc/src/main/java/com/linecorp/armeria/server/grpc/DeferredListener.java

```diff
@@ -75,104 +74,93 @@
+                    final List<Consumer<Listener<I>>> pendingTasks = this.pendingTasks;
+                    if (pendingTasks == null) {
+                        break;
+                    }
+
+                    this.pendingTasks = null;
+                    try {
+                        for (Consumer<Listener<I>> task : pendingTasks) {
+                            task.accept(delegate);
+                        }
+                    } catch (Throwable ex) {
+                        callClosed = true;
+                        armeriaServerCall.close(ex);
+                        return null;
+                    }
                 }
-            } catch (Throwable ex) {
-                callClosed = true;
-                armeriaServerCall.close(ex);
-                return null;
+            } finally {
+                shouldBePending = false;
             }
-            //noinspection unchecked
-            pendingTasks = (List<Consumer<Listener<I>>>) NOOP_TASKS;
             return null;
         }, sequentialExecutor());
     }
 
     @Override
     public void onMessage(I message) {
-        if (callClosed) {
-            return;
-        }
-
-        if (pendingTasks == NOOP_TASKS) {
-            // listenerFuture has completed successfully. No race with the listenerFuture's callback.
-            delegate.onMessage(message);
-        } else {
-            maybeAddPendingTask(listener -> listener.onMessage(message));
-        }
+        maybeAddPendingTask(listener -> listener.onMessage(message));
     }
 
     @Override
     public void onHalfClose() {
-        if (callClosed) {
-            return;
-        }
-
-        if (pendingTasks == NOOP_TASKS) {
-            delegate.onHalfClose();
-        } else {
-            maybeAddPendingTask(Listener::onHalfClose);
-        }
+        maybeAddPendingTask(Listener::onHalfClose);
     }
 
     @Override
     public void onCancel() {
-        if (callClosed) {
-            return;
-        }
-
-        if (pendingTasks == NOOP_TASKS) {
-            delegate.onCancel();
-        } else {
-            maybeAddPendingTask(Listener::onCancel);
-        }
+        maybeAddPendingTask(Listener::onCancel);
     }
 
     @Override
     public void onComplete() {
-        if (callClosed) {
-            return;
-        }
-
-        if (pendingTasks == NOOP_TASKS) {
-            delegate.onComplete();
-        } else {
-            maybeAddPendingTask(Listener::onComplete);
-        }
+        maybeAddPendingTask(Listener::onComplete);
     }
 
     @Override
     public void onReady() {
+        maybeAddPendingTask(Listener::onReady);
+    }
+
+    private void maybeAddPendingTask(Consumer<ServerCall.Listener<I>> task) {
         if (callClosed) {
             return;
         }
 
-        if (pendingTasks == NOOP_TASKS) {
-            delegate.onReady();
-        } else {
-            maybeAddPendingTask(Listener::onReady);
+        if (!shouldBePending()) {
+            task.accept(delegate);
```

Comment:
`delegate` may be null event if `!shouldBePending()` is true because of the [reordering](https://www.logicbig.com/tutorials/core-java-tutorial/java-multi-threading/thread-reordering.html) problem.
Could we check if `delegate` is not null as well?

```suggestion
        if (!shouldBePending() && delegate != null) {
            task.accept(delegate);
```

Reply:
`!shouldBePending()` is true only after exiting the loop. Can it be null in the loop?